numpy/__init__.pyi

```diff
@@ -929,51 +930,51 @@ class dtype(Generic[_DTypeScalar_co]):
     @overload
     def __new__(cls, dtype: Type[bytes], align: bool = ..., copy: bool = ...) -> dtype[bytes_]: ...
 
-    # `unsignedinteger` string-based representations
+    # `unsignedinteger` string-based representations and ctypes
     @overload
-    def __new__(cls, dtype: _UInt8Codes, align: bool = ..., copy: bool = ...) -> dtype[uint8]: ...
+    def __new__(cls, dtype: _UInt8Codes | Type[ct.c_uint8], align: bool = ..., copy: bool = ...) -> dtype[uint8]: ...
     @overload
-    def __new__(cls, dtype: _UInt16Codes, align: bool = ..., copy: bool = ...) -> dtype[uint16]: ...
+    def __new__(cls, dtype: _UInt16Codes | Type[ct.c_uint16], align: bool = ..., copy: bool = ...) -> dtype[uint16]: ...
     @overload
-    def __new__(cls, dtype: _UInt32Codes, align: bool = ..., copy: bool = ...) -> dtype[uint32]: ...
+    def __new__(cls, dtype: _UInt32Codes | Type[ct.c_uint32], align: bool = ..., copy: bool = ...) -> dtype[uint32]: ...
     @overload
-    def __new__(cls, dtype: _UInt64Codes, align: bool = ..., copy: bool = ...) -> dtype[uint64]: ...
+    def __new__(cls, dtype: _UInt64Codes | Type[ct.c_uint64], align: bool = ..., copy: bool = ...) -> dtype[uint64]: ...
     @overload
-    def __new__(cls, dtype: _UByteCodes, align: bool = ..., copy: bool = ...) -> dtype[ubyte]: ...
+    def __new__(cls, dtype: _UByteCodes | Type[ct.c_ubyte], align: bool = ..., copy: bool = ...) -> dtype[ubyte]: ...
     @overload
-    def __new__(cls, dtype: _UShortCodes, align: bool = ..., copy: bool = ...) -> dtype[ushort]: ...
+    def __new__(cls, dtype: _UShortCodes | Type[ct.c_ushort], align: bool = ..., copy: bool = ...) -> dtype[ushort]: ...
     @overload
-    def __new__(cls, dtype: _UIntCCodes, align: bool = ..., copy: bool = ...) -> dtype[uintc]: ...
+    def __new__(cls, dtype: _UIntCCodes | Type[ct.c_uint], align: bool = ..., copy: bool = ...) -> dtype[uintc]: ...
     @overload
-    def __new__(cls, dtype: _UIntPCodes, align: bool = ..., copy: bool = ...) -> dtype[uintp]: ...
+    def __new__(cls, dtype: _UIntPCodes | Type[ct.c_void_p] | Type[ct.c_size_t], align: bool = ..., copy: bool = ...) -> dtype[uintp]: ...
```

Comment:
To verify: are there any situations where `ctypes.c_void_p` and `ctypes.c_size_t` don't map to `np.uintp`?

Reply:
Yes there are, platforms where `uint_ptr_t != size_t`.

Numpy is probably already broken on those platforms though.

Reply:
> Numpy is probably already broken on those platforms though.

Based on this I feel it would be adequate to add a comment for these cases and call it a day (https://github.com/numpy/numpy/pull/19468/commits/e7cd1e78ce728e3b745cea5a34c7fdea0f5c378e).

Reply:
Just to check... Is this actually necessary? `ct.c_size_t` is just a type alias for one of the other C types, unlike `ctypes.c_void_p` which is actually a different type.

Is `uintp` considered a distinct type by mypy, or is it also treated as an alias of another type?

Reply:
> Just to check... Is this actually necessary?

Unfortunately it is, as the various `ctypes`aliases are all annotated as distinct types in its [stub file](https://github.com/python/typeshed/blob/0cd4ee39a2fbe17001b1a98929158ffc0b11f664/stdlib/ctypes/__init__.pyi#L227-L250).
Mypy does have a [builtin plugin](https://github.com/python/mypy/blob/master/mypy/plugins/ctypes.py) for dealing with `ctypes.Array`, but there doesn't seem to be anything
for dealing with type aliases (or the lack thereof).

> Is `uintp` considered a distinct type by mypy,  or is it also treated as an alias of another type?

No, on numpy's side all `np.unsignedinteger` subclasses are considered interchangeable (as long as 
they're of the size). So on a 64-bit platform `np.uint64` and `np.uintp` are equivalent as far as mypy 
is concerned. 

So with this in mind, I'd say that the `np.generic` subclasses are currently a bit easier to use with typing 
than `ctypes._SimpleCData`.

https://github.com/numpy/numpy/blob/89c80ba606f4346f8df2a31cfcc0e967045a68ed/numpy/__init__.pyi#L3287-L3298